# Наивный байесовский классификатор

In [1]:
import numpy as np
import matplotlib.pyplot as plt
%pylab inline

Populating the interactive namespace from numpy and matplotlib


In [2]:
from sklearn import datasets

In [3]:
digits = datasets.load_digits()
breast_cancer = datasets.load_breast_cancer()

In [4]:
from pandas import DataFrame

In [5]:
digits_frame = DataFrame(digits.data)
digits_frame['target'] = digits.target

In [6]:
digits_frame.head()

0  1  2   3   4   5  6  7  8  9   ...    55  56  57  58  59  60  61  62  \
0  0  0  5  13   9   1  0  0  0  0   ...     0   0   0   6  13  10   0   0   
1  0  0  0  12  13   5  0  0  0  0   ...     0   0   0   0  11  16  10   0   
2  0  0  0   4  15  12  0  0  0  0   ...     0   0   0   0   3  11  16   9   
3  0  0  7  15  13   1  0  0  0  8   ...     0   0   0   7  13  13   9   0   
4  0  0  0   1  11   0  0  0  0  0   ...     0   0   0   0   2  16   4   0   

   63  target  
0   0       0  
1   0       1  
2   0       2  
3   0       3  
4   0       4  

[5 rows x 65 columns]

In [7]:
breast_cancer_frame = DataFrame(breast_cancer.data)
breast_cancer_frame.columns = breast_cancer.feature_names
breast_cancer_frame['target'] = breast_cancer.target

In [8]:
breast_cancer_frame.head()

mean radius  mean texture  mean perimeter  mean area  mean smoothness  \
0        17.99         10.38          122.80     1001.0          0.11840   
1        20.57         17.77          132.90     1326.0          0.08474   
2        19.69         21.25          130.00     1203.0          0.10960   
3        11.42         20.38           77.58      386.1          0.14250   
4        20.29         14.34          135.10     1297.0          0.10030   

   mean compactness  mean concavity  mean concave points  mean symmetry  \
0           0.27760          0.3001              0.14710         0.2419   
1           0.07864          0.0869              0.07017         0.1812   
2           0.15990          0.1974              0.12790         0.2069   
3           0.28390          0.2414              0.10520         0.2597   
4           0.13280          0.1980              0.10430         0.1809   

   mean fractal dimension   ...    worst texture  worst perimeter  worst area  \
0                 0.07871   ...            17.33           184.60      2019.0   
1                 0.05667   ...            23.41           158.80      1956.0   
2                 0.05999   ...            25.53           152.50      1709.0   
3                 0.09744   ...            26.50            98.87       567.7   
4                 0.05883   ...            16.67           152.20      1575.0   

   worst smoothness  worst compactness  worst concavity  worst concave points  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   worst symmetry  worst fractal dimension  target  
0          0.4601                  0.11890       0  
1          0.2750                  0.08902       0  
2          0.3613                  0.08758       0  
3          0.6638                  0.17300       0  
4          0.2364                  0.07678       0  

[5 rows x 31 columns]

In [9]:
from sklearn import naive_bayes
from sklearn import model_selection

In [10]:
estimators = np.array([naive_bayes.BernoulliNB(), naive_bayes.MultinomialNB(), naive_bayes.GaussianNB()])
estimators_str = np.array(["BernoulliNB", "MultinomialNB", "GaussianNB"])
for i in range(estimators.size):
    print estimators_str[i]
    print "Digits: " + str(model_selection.cross_val_score(estimators[i], digits.data, digits.target).mean())
    print "Breast_cancer: " + str(model_selection.cross_val_score(estimators[i], breast_cancer.data, 
                                                                        breast_cancer.target).mean())

BernoulliNB
Digits: 0.825823650778
Breast_cancer: 0.627420402859
MultinomialNB
Digits: 0.870877148974
Breast_cancer: 0.894579040193
GaussianNB
Digits: 0.818600380355
Breast_cancer: 0.936749280609


Результат эксперимента согласуются с ожиданиями: вещественные признаки лучше описываются непрерывным распределением, а целые признаки - дискретным.

### Ответы на вопросы

1. Максимальное качество на датасете breast_cancer(0.936749280609) получилось при выборе наивного байесовского классификатора GaussianNB
2. Максимальное качество классификации на датасете digits(0.870877148974) - при выборе MultinomialNB
3. По результатам эксперимента верны утверждения (c) и (d): мультиномиальное распределение показало себя лучше на выборке с целыми неотрицательными значениями признаков - датасет digits, а на вещественных признаках - датасет breast_canser - лучше сработало нормальное распределение.